src/sys/pthread.rs

```diff
@@ -34,10 +39,351 @@ pub fn pthread_kill<T>(thread: Pthread, signal: T) -> Result<()>
+/// const DELTA: Duration = Duration::from_millis(100);
+/// # fn main() -> nix::Result<()> {
+/// let mutex = Mutex::new(None)?;
+/// let mutex_arc = Arc::new(mutex);
+/// let mutex_clone = mutex_arc.clone();
+/// let instant = Instant::now();
+/// let handle = spawn(move || -> nix::Result<(),> {
+///     let guard = mutex_clone.lock()?;
+///     sleep(TIMEOUT);
+///     guard.try_unlock()?;
+///     Ok(())
+/// });
+/// sleep(DELTA);
+/// let guard = mutex_arc.lock()?;
+/// assert!(instant.elapsed() > TIMEOUT && instant.elapsed() < TIMEOUT + DELTA);
+/// assert_eq!(handle.join().unwrap(), Ok(()));
+/// # Ok(())
+/// # }
+/// ```
+/// 
+/// ### Multi-process
+/// 
+/// ```
+/// # use std::{
+/// #   sync::Arc,
+/// #   time::{Instant, Duration},
+/// #   thread::{sleep, spawn},
+/// #   mem::{size_of, MaybeUninit},
+/// #   num::NonZeroUsize,
+/// #   os::unix::io::OwnedFd
+/// # };
+/// # use nix::{
+/// #   sys::{pthread::{Mutex, MutexAttr}, mman::{mmap, MapFlags, ProtFlags}},
+/// #   unistd::{fork,ForkResult},
+/// # };
+/// const TIMEOUT: Duration = Duration::from_millis(500);
+/// const DELTA: Duration = Duration::from_millis(100);
+/// # fn main() -> nix::Result<()> {
+/// let shared_memory = unsafe { mmap::<OwnedFd>(
+///     None,
+///     NonZeroUsize::new_unchecked(size_of::<Mutex>()),
+///     ProtFlags::PROT_WRITE | ProtFlags::PROT_READ,
+///     MapFlags::MAP_SHARED | MapFlags::MAP_ANONYMOUS,
+///     None,
+///     0
+/// )? };
+/// 
+/// let mutex = unsafe {
+///     let mutex_ptr = shared_memory.cast::<Mutex>();
+/// 
+///     // A mutex must be initialized.
+///     // By default mutex's are process private, so we also need to initialize with the
+///     // `MutexAttr` with shared.
+///     let mut mutex_attr = MutexAttr::new()?;
+///     mutex_attr.set_shared(true)?;
+///     Mutex::init(mutex_ptr, Some(mutex_attr))?;
+/// 
+///     &*mutex_ptr
+/// };
+/// 
+/// match unsafe { fork()? } {
+///     ForkResult::Parent { child } => {
+///         let guard = mutex.lock()?;
+///         sleep(TIMEOUT);
+///         guard.try_unlock();
+///         // Wait for child process to exit
+///         unsafe {
+///             assert_eq!(libc::waitpid(child.as_raw(), std::ptr::null_mut(),0), child.as_raw());
+///         }
+///     },
+///     ForkResult::Child => {
+///         let now = Instant::now();
+///         sleep(DELTA);
+///         mutex.lock()?;
+///         assert!(now.elapsed() > TIMEOUT && now.elapsed() < TIMEOUT + DELTA);
+///     }
+/// }
+/// 
+/// # Ok(())
+/// # }
+/// ```
+#[cfg(target_os = "linux")]
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct Mutex(UnsafeCell<libc::pthread_mutex_t>);
+
+#[cfg(target_os = "linux")]
+impl Mutex {
+    /// Wraps [`libc::pthread_mutex_init`].
+    /// 
+    /// # Safety
+    /// 
+    /// Attempting to initialize an already initialized mutex results in undefined behavior.
+    pub unsafe fn init(mutex: *mut Mutex, attr: Option<MutexAttr>) -> Result<()> {
```

Comment:
What I had in mind was passing the raw pointer directly to libc.  I'm not sure that taking a `*mut Mutex` pointer is sufficient, if the target is not initialized.  Can you find any other examples of initializing objects in shared memory?  Would it be better to construct the mutex elsewhere, and then move it into shared memory?  For that matter, the shared memory will only be useful if it can be accessed from more than one process.  How do you plan for the second process to verify that the mutex is initialized?

Reply:
> How do you plan for the second process to verify that the mutex is initialized?

I'm not sure there is away to do this in this scope. In practice I'd initialize the mutex before forking, but this cannot be captured here.

I think there is only so much this PR can cover, I would say we shouldn't let perfect be the enemy of good.